### PR TITLE
Add logic to scrape github content

### LINKED ISSUE #5: add mastering bitcoin to scraper
https://github.com/bitcoinbook/bitcoinbook/

### PR DESCRIPTION
**Pull request by nassersaazi**
This PR adds the scrapers to close #5 #6 #10 #11 #14 #15

### CODE REVIEW

File: githubcontent/main.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    documents = []
+    for url in urls:
+        is_bitcoin_url = re.search('bitcoinbook', url)
+        data = requests.get(url).text
+        soup = BeautifulSoup(data,'html.parser')
+        document = {}
+        title = soup.find('h2', dir='auto').get_text()
+        body = soup.find('div',id = 'readme').get_text()
+        body_type = "md"
+        author = "Andreas Antonopoulos"
```
Comment (adamjonas) [inline githubcontent/main.py]: there are other authors for LN

File: githubcontent/main.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        id = "bitcoinbook-chapter-" + chapter_number if is_bitcoin_url else "lnbook-chapter-" + chapter_number 
+        domain = "https://github.com"
+        url = url
+        created_at = "2022-11-15" if is_bitcoin_url else "2023-04-22"# date of most recent commit
+
+        document.update({
+            "title": title,
+            "body": body,
+            "body_type": body_type,
+            "author": author,
```
Comment (adamjonas) [inline githubcontent/main.py]: believe it should be plaural - authors

File: githubcontent/main.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    documents = []
+    for url in urls:
+        is_bitcoin_url = re.search('bitcoinbook', url)
+        data = requests.get(url).text
+        soup = BeautifulSoup(data,'html.parser')
+        document = {}
+        title = soup.find('h2', dir='auto').get_text()
+        body = soup.find('div',id = 'readme').get_text()
+        body_type = "md"
```
Comment (adamjonas) [inline githubcontent/main.py]: it's not MD though

File: githubcontent/main.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        author = "Andreas Antonopoulos"
+        chapter_number = ''.join(re.findall(r'\d+', url))
+        id = "bitcoinbook-chapter-" + chapter_number if is_bitcoin_url else "lnbook-chapter-" + chapter_number 
+        domain = "https://github.com"
+        url = url
+        created_at = "2022-11-15" if is_bitcoin_url else "2023-04-22"# date of most recent commit
+
+        document.update({
+            "title": title,
+            "body": body,
```
Comment (adamjonas) [inline githubcontent/main.py]: I'd like to add a field that is body_formatted as well to reflect https://github.com/bitcoinsearch/scraper/issues/18

Comment (nassersaazi) [reply githubcontent/main.py]: @adamjonas should this be a boolean or some kind of enumeration?

File: githubcontent/main.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    for url in urls:
+        is_bitcoin_url = re.search('bitcoinbook', url)
+        data = requests.get(url).text
+        soup = BeautifulSoup(data,'html.parser')
+        document = {}
+        title = soup.find('h2', dir='auto').get_text()
+        body = soup.find('div',id = 'readme').get_text()
+        body_type = "raw"
+        authors = ["Andreas M. Antonopoulos"] if is_bitcoin_url else ["Andreas M. Antonopoulos","Olaoluwa Osuntokun","Rene Pickhardt"]
+        id = str(uuid.uuid4())
```
Comment (adamjonas) [inline githubcontent/main.py]: I'd suggest you stick with the previously used semantic convention for IDs

https://github.com/bitcoinsearch/scraper/blob/master/bitcoinops/main.js#L105

File: githubcontent/main.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    documents = []
+    for url in chapter_links:
+        data = requests.get(url).text
+        soup = BeautifulSoup(data,'html.parser')
+        document = {}
+        title = soup.find('h2', dir='auto').get_text()
+        body = soup.find('article').get_text()
+        body_type = "raw"
+        author = "Jimmy Song"
+        id = str(uuid.uuid4())
```
Comment (adamjonas) [inline githubcontent/main.py]: same as https://github.com/bitcoinsearch/scraper/pull/28/files#r1205810644

File: githubcontent/main.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    documents = []
+    for url in urls:
+        is_bitcoin_url = re.search('bitcoinbook', url)
+        data = requests.get(url).text
+        soup = BeautifulSoup(data,'html.parser')
+        document = {}
+        title = soup.find('h2', dir='auto').get_text()
+        body = soup.find('div',id = 'readme').get_text()
+        body_type = "raw"
+        authors = ["Andreas M. Antonopoulos"] if is_bitcoin_url else ["Andreas M. Antonopoulos","Olaoluwa Osuntokun","Rene Pickhardt"]
```
Comment (adamjonas) [inline githubcontent/main.py]:
```suggestion
        authors = ["Andreas Antonopoulos"] if is_bitcoin_url else ["Andreas Antonopoulos","Olaoluwa Osuntokun","Rene Pickhardt"]
```

File: githubcontent/main.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+    documents = []
+    for url in urls:
+        is_bitcoin_url = re.search('bitcoinbook', url)
+        data = requests.get(url).text
+        soup = BeautifulSoup(data,'html.parser')
+        document = {}
+        title = soup.find('h2', dir='auto').get_text()
+        body = soup.find('div',id = 'readme').get_text()
+        body_type = "raw"
```
Comment (adamjonas) [inline githubcontent/main.py]: They aren't raw though. They are asciidoc. We are throwing away the titles and headers but not persisting them.

File: githubcontent/main.py (hunk 4 above)
Comment (adamjonas) [reply githubcontent/main.py]: no, this is likely a formatted body similar to how things are parsed in our markdown parsers:

https://github.com/bitcoinsearch/scraper/blob/master/bitcoinops/main.js#L94

File: githubcontent/utils.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+    def get_details(self, details: list):
+        result_dict = {}
+
+        for item in details:
+            if ': ' in item:
+                key, value = item.split(': ', 1)
+                result_dict[key.strip()] = value.strip()
+            else:
+                print(f"Ignoring item: {item}")
```
Comment (adamjonas) [inline githubcontent/utils.py]: what does it mean when so many are ignored for these BIPs?

```
Ignoring item:            OPL
bips-0f50ae2e-39ce-4c65-b8bc-111bab20404f
Ignoring item:           Luke Dashjr <luke+bip@dashjr.org>
Ignoring item:            CC0-1.0
bips-b13926d7-32aa-48c3-8498-51140e953690
Ignoring item:           Peter Todd <pete@petertodd.org>
Ignoring item:           Greg Maxwell <greg@xiph.org>
Ignoring item:           Rusty Russell <rusty@rustcorp.com.au>
```

File: githubcontent/bips.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    data = requests.get(base_url).text
+    soup = BeautifulSoup(data,'html.parser')
+    table = soup.find('table')
+    links = table.find_all('a')
+    urls = [link['href'] for link in links]
+    for url in urls:
+        bip_url = domain + url
+        data = requests.get(bip_url).text
+        soup = BeautifulSoup(data,'html.parser')
+        body = soup.find('article').get_text()
```
Comment (adamjonas) [inline githubcontent/bips.py]: so this is nice for the raw text but we need the formatted text as well like here: https://github.com/bitcoinsearch/scraper/blob/master/bitcoinops/main.js#L94

We want both. We want the raw text as well as the formatted text. Without the formatted text we don't get the headers standing out from the paragraph text.

File: githubcontent/blips.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        soup = BeautifulSoup(data,'html.parser')
+        body = soup.find('article').get_text()
+        details = soup.find(class_='notranslate').get_text()
+        details = details.split("\n")
+        blip_info = scraper.get_details(details[:-1])
+        document = {}
+        document.update({
+            "title": blip_info.get("Title"),
+            "body": body,
+            "body_type": "md",
```
Comment (adamjonas) [inline githubcontent/blips.py]:
```suggestion
            "body_type": "markdown",
```

File: githubcontent/bolts.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+    base_url = 'https://github.com/lightning/bolts/blob/master/{}.md'
+    print("Getting links for lightning bolts")
+    bolt_links = scraper.get_github_urls(base_url,bolts)
+    for link in bolt_links:
+        data = requests.get(link).text
+        soup = BeautifulSoup(data,'html.parser')
+        title = soup.find('h1',dir='auto').get_text()
+        body = soup.find('article').get_text()
+        body_type = "md"
```
Comment (adamjonas) [inline githubcontent/bolts.py]:
```suggestion
        body_type = "markdown"
```

File: githubcontent/lndocs.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        data = requests.get(doc_url).text
+        soup = BeautifulSoup(data,'html.parser')
+        document = {}
+        article = soup.find('article')
+        body = article.get_text()
+        title = article.find('h1').get_text()
+        document.update({
+            "title": title,
+            "body": body,
+            "body_type": "md",
```
Comment (adamjonas) [inline githubcontent/lndocs.py]:
```suggestion
            "body_type": "markdown",
```

File: githubcontent/utils.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    def parse_aantonop_books(self,urls):
+
+        documents = []
+        for url in urls:
+            is_bitcoin_url = re.search('bitcoinbook', url)
+            data = requests.get(url).text
+            soup = BeautifulSoup(data,'html.parser')
+            document = {}
+            title = soup.find('h2', dir='auto').get_text()
+            body = soup.find('div',id = 'readme').get_text()
```
Comment (adamjonas) [inline githubcontent/utils.py]: same comment as above about not capturing headers

File: githubcontent/blips.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+            "authors": [blip_info.get("Author")],
```
Comment (adamjonas) [inline githubcontent/blips.py]: this pulls the name and email e.g. `"Ryan Gentry <ryan@lightning.engineering>"`, we want just the name `"Ryan Gentry"`

File: githubcontent/bolts.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        authors = []
```
Comment (adamjonas) [inline githubcontent/bolts.py]: I don't like leaving this empty. Maybe something like this?
```suggestion
        authors = ["Spec"]
```

File: githubcontent/utils.py [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            body_type = "asciidoc"
+            authors = ["Andreas Antonopoulos"] if is_bitcoin_url else ["Andreas Antonopoulos","Olaoluwa Osuntokun","Rene Pickhardt"]
+            id = 'masteringbitcoin' + str(uuid.uuid4()) if is_bitcoin_url else 'masteringln' + str(uuid.uuid4())
+            domain = "https://github.com"
+            url = url
+            created_at = "2022-11-15" if is_bitcoin_url else "2023-04-22"# date of most recent commit
```
Comment (adamjonas) [inline githubcontent/utils.py]: that's the last commit on the repo, not the last commit on the resource right? (low priority)

File: githubcontent/bips.py [hunk 18]
```diff
@@ -0,0 +1,54 @@
+from bs4 import BeautifulSoup
```
Comment (adamjonas) [inline githubcontent/bips.py]: we need this added to requirements.txt. I've pushed a commit that moves it out of the StackExchange folder to the root.

**Comment by adamjonas** [comment]
The OP needs updating as there is no `main.py`

File: githubcontent/utils.py (hunk 9 above)
Comment (adamjonas) [reply githubcontent/utils.py]: ping on this ^^

File: scrapybot/scrapybot/spiders/bolts.py [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+        ),
+    )
+
+    def parse_item(self, response):
+        item = {}
+
+        body_to_be_parsed = response.xpath("//article").get()
+        item["id"] = "bolts-" + str(uuid.uuid4())
+        item["title"] = response.xpath('//h1[@dir="auto"]/text()').get()
+        item["body"] = strip_tags(body_to_be_parsed)
```
Comment (adamjonas) [inline scrapybot/scrapybot/spiders/bolts.py:27]: so for the asciidoc and markdown files, it seems like you stripped the headers/tags and uploaded the raw text or did you capture the headers in another way?

File: scrapybot/scrapybot/spiders/rusty.py [hunk 20]
```diff
@@ -0,0 +1,31 @@
+import re
+import uuid
+from .utils import strip_tags
+from scrapy.linkextractors import LinkExtractor
+from scrapy.spiders import CrawlSpider, Rule
+
+
+class RustySpider(CrawlSpider):
+    name = "rusty"
+    allowed_domains = ["rusty.ozlabs.org"]
```
Comment (adamjonas) [inline scrapybot/scrapybot/spiders/rusty.py]: I don't want all of these, some are quite unrelated to bitcoin: https://rusty.ozlabs.org/?paged=11

File: githubcontent/utils.py (hunk 9 above)
Comment (nassersaazi) [reply githubcontent/utils.py]: this has been resolved in the latest PR in [this commit](https://github.com/bitcoinsearch/scraper/pull/30/commits/839079770e15ffd4cd5d7d407b8adc8033b7f4c3#diff-c2ea54204bb3c234d87aad9d28667980744deb328765ab476427840de6db0573)

File: githubcontent/utils.py (hunk 17 above)
Comment (nassersaazi) [reply githubcontent/utils.py]: yes

File: scrapybot/scrapybot/spiders/bolts.py (hunk 19 above)
Comment (nassersaazi) [reply scrapybot/scrapybot/spiders/bolts.py:27]: in the [new setup](https://github.com/bitcoinsearch/scraper/pull/30) i capture the immediate html tags as well as the raw text

**Pull request closed without merging**